### PR TITLE
Fix Hyperlink/Travis Issues

### PR DESCRIPTION
**Pull request by skipjack**
Reverted 8281949b400 so we can test and try to resolve the `hyperlink` issue on Travis' network. I think there still may be one link failure due to a package's README that we may have to resolve before we can fix the core issue.

Related to Munter/hyperlink#121

### CODE REVIEW

**Comment by skipjack** [comment]
@Munter I figured out the issue here and it's actually more a problem with our customizations to the `hyperlink` call than `hyperlink` itself. See the [check-links.js]() script...

The issue is that we don't output anything until `hyperlink` completes which allows us to filter through the failures to exclude checks on specific elements/urls. It would be great if there was an option in `hyperlink` for element-based exclusion, however I'm guessing we could also update the script to output to the console while the script is running to avoid no output for such a long period of time. The benefit to native exclusion is that the script could actually just avoid testing certain links altogether thus improving performance and making this script unnecessary.

I'm not too familiar with the `tap` ecosystem but if you have any advice, that'd be much appreciated. I see that there is an `--exclude` option but only for URLs, and unfortunately in one of those two cases (`class="support__[^"]*"`) there's no consistent pattern for the URLs. Interested in your thoughts on the best approach so we can get this resolved.

**Comment by Munter** [comment]
I think the solution might be to use an appropriate tap-reporter or write your own with a filter. Most tap reporters I have seen have continuous reporting of the tap stream. Alternatively you could use tee to pipe the tap output to a file while still logging all the activity to the console, then post-process the tap output from the file

**Comment by skipjack** [comment]
> Most tap reporters I have seen have continuous reporting of the tap stream.

Yeah so I updated our script to check and output line by line, instead of waiting for completion, but unfortunately as you can see the build still fails for some reason. There must be one link check that's hanging for over 10 min. I'm thinking a quick hack would just be to use a `setInterval` to log a `...` string every few min to prevent it from completely failing out however that's starting to feel pretty desperate.

**Comment by Munter** [comment]
Looks like there is not default timeout for `teepee`, which is the http client assetgraph uses. I'll see if I can set a relevant default and maybe expose the acceptable timeout as a cli option.

image

**Comment by skipjack** [comment]
@Munter somehow it's still hanging on the link following this line:

`loading build/1ebd0482aadade65f20ec178219fe012.woff2`

(this is with the patch fix I made for your `fix/defaultTimeout` branch)

**Comment by skipjack** [comment]
@Munter btw I'd be more than happy to add you as a contributor if you're interested. It might be easier for you to debug this branch/pr that way.

**Comment by Munter** [comment]
@skipjack yeah, that might be useful. I'm not sure when I'll have the time to do a deep dive on this, but when I do it should speed up my feedback loop

**Comment by skipjack** [comment]
> yeah, that might be useful. I'm not sure when I'll have the time to do a deep dive on this, but when I do it should speed up my feedback loop

No worries, just sent you the collaborator invite. For now I'm just running `hyperlink` locally on bigger PRs that might cause link failures. Let me know if there's anything else I can do to help.

**Comment by skipjack** [comment]
@pierreneter would you be interested in submitting a PR to the `hyperlink` repo to resolve Munter/hyperlink#121 and thus fix this issue? We basically need to somehow shut down any long-running requests, I'm sure @Munter would help if you have ideas.

I've been pushing hard to get the site's information architecture sorted out and sections cleaned up which I think will be pretty complete by the end of this week. I have a plan on how to start working at #1525 and #1400 starting next week (12/24-ish), assuming all goes to plan. Hoping to review that with you and @jeremenichelli if you're both still interested and willing to help.

That said, it would be good to get this resolved before we dig into that though as `hyperlink` is a great way for us to make sure we aren't breaking things when making major changes. If you, or anyone else, would be willing to tackle this I'd be very appreciative 🙏  .

**Comment by pierreneter** [comment]
Ok. I will try

**Comment by jeremenichelli** [comment]
@skipjack I'm in for the next steps towards this. Just ping me on the right discussions, I haven't been in Slack too much the last weeks because of personal stuff.

I think that moving to another static generator like jekyll, hugo, gatsby, etc. is worth a hangout to see what's best given the complexity of a hypothetical migration and the amount of work that it will mean. I personally would love that to happen :)

**Comment by pierreneter** [comment]
@skipjack I think @Munter just did it. We need update `hyperlink` version to try it work or not. But, what about a handling for `timeout`?

**Comment by pierreneter** [comment]
About #1525, #1400. The same with @jeremenichelli, I vote for `gatsby` but we need more discussion

**Comment by pierreneter** [comment]
See logs at https://travis-ci.org/webpack/webpack.js.org/builds/318201104 It looks like set `timeout` for `teepee` that does not solve our problem

**Comment by pierreneter** [comment]
It's stuck at a font file (`woff2` - is not plain text). It's local file and not a problem with the request. `PNG` is also not pure text, but it's passed quickly. I think we just need to check local files (it aren't plain text) exist or not, not need to load it. @Munter what do you think?
P/s: @skipjack the small problem, we need check URLs that are external URL link like: `"medium.com/@ridermansb"` (https://travis-ci.org/webpack/webpack.js.org/builds/318427547#L2228) with missing URI schema and fix it.

**Comment by skipjack** [comment]
@pierreneter I'm actually not sure it's hanging on the font file even though it's the last log line:

``` bash
loading build/314bbcd238d458622bbf32427346774f.woff
```

I think it may be loading the next one but failing before it makes to logging it out (though I could be wrong). I'm pretty sure it's one of the sponsor/backer links. I think we have to debug more at the `hyperlink` level -- the issue is that whatever the case may be in terms of fixing URLs or network issues, `hyperlink` abort a request after a minute or two (or some user provided time period).

**Comment by pierreneter** [comment]
@skipjack @Munter added abort a request after 30 seconds timeout (https://github.com/Munter/hyperlink/commit/a0bb4b30c20b44066527f04210e66c85cfeaf17e) but I don't know it works or not.

**Comment by skipjack** [comment]
> Just ping me on the right discussions, I haven't been in Slack too much the last weeks because of personal stuff.

@jeremenichelli Cool, no problem. Yeah I was thinking maybe you could take the lead on the refactoring the `fetch` scripts (i.e. the ones that populate external docs not local to the site). As we discussed a while back, some prefix like `_` could be used on those files for a simpler `.gitignore` and general structure.

> I think that moving to another static generator like jekyll, hugo, gatsby, etc. is worth a hangout to see what's best given the complexity of a hypothetical migration and the amount of work that it will mean. I personally would love that to happen :)
> 
> I vote for gatsby but we need more discussion

@pierreneter @jeremenichelli yeah if we're going to switch to another SSG, I think Gatsby would be the best alternative. @pierreneter I did finally track down the `StaticSiteGeneratorPlugin` [approach](https://jaketrent.com/post/react-routing-static-site-browser/) which would yield something very similar to what Gatsby outputs but with pure webpack. Personally I think this approach would be a bit nicer as we'd retain full control over the process and in some ways it would kinda poetic to build the webpack site with just webpack rather than "endorse" a generator when there are already a lot out there.

How about this, I'll throw together a branch within the next two weeks to demo what I mean on the pure webpack approach. If that feels clunky, we can take a stab at porting to Gatsby. If not, maybe we continue down the pure webpack path and see where it takes us. Either of those approaches would leave us in a better spot than we are now and would make the tasks in #1525 and #1400 easier to tackle. The question in my mind is more, do we port to Gatsby and let it handle everything for us behind the scenes or can we build a relatively simple webpack config that yields the same thing but leaves us with more control if needed (btw, Gatsby utilizes the `StaticSiteGeneratorPlugin` under the hood).

Let's not clutter this thread more (my bad 😁 ) but I really will ping you guys soon to finish discussing infrastructure.

**Comment by skipjack** [comment]
> added abort a request after 30 seconds timeout (Munter/hyperlink@a0bb4b3) but I don't know it works or not.

@pierreneter could you use a [git dependency](https://docs.npmjs.com/files/package.json#git-urls-as-dependencies) in the `package.json` on this branch to test it out?

**Comment by skipjack** [comment]
@pierreneter actually I think that commit was the one from a while ago that we tested out but didn't work. Feel free to try again though.

**Comment by Munter** [comment]
I see several possible causes of error here. Might be some request in a promise that throws unexpectedly and is not caught. Or some crazy long download or an actual timeout. I was considering pushing everything that initiates a request in some sort of request tracking array and then removing it again when the request has resolved. Then start logging out that arrays content periodically , which means if there is a 10 minute hang that stops the CI server, we should see the unhandled connections logged in the console. That would let us know what resources cause this and possibly how to recreate it locally to implement a test against it

**Comment by pierreneter** [comment]
I added more logs to `request`s of my custom `hyperlink`. CI test's result said that `request`s for try connect and http status has not been used. It seems that it is stuck at the file loading process. So many file with so many link

**Comment by skipjack** [comment]
@Munter any idea why the `teepee` timeout isn't working? I poked around a little, which was a bit tough as the `teepee` property isn't mentioned in `assetgraph`s readme, but managed to find the [related code][1] and then poked around in `teepee`s source to find the [`timeout` related][2] bits. 

I was initially thinking we could do the same thing with a `setTimeout` within `hyperlink` but it seems like all requests are basically handled behind the scenes by `assetgraph` and `teepee`? Should our next steps just be to `npm link` the `hyperlink` pack and then `npm link` `assetgraph` and/or `teepee` so we can debug the whole stack?

> I was considering pushing everything that initiates a request in some sort of request tracking array and then removing it again when the request has resolved. Then start logging out that arrays content periodically , which means if there is a 10 minute hang that stops the CI server, we should see the unhandled connections logged in the console.

Where can we get access the requests? Somewhere from within the `assetgraph` (`ag`) instance? If we can do that maybe there's some way we can invoke `teepee` to abort longstanding requests (as a short-term fix without having to `npm link` the whole stack?

[1]: https://github.com/assetgraph/assetgraph/blob/e137bd81407a1632a4441956ad3aa012ec1eb745/lib/AssetGraph.js#L95-L100
[2]: https://github.com/One-com/teepee/blob/master/lib/Teepee.js#L629-L636

**Comment by pierreneter** [comment]
`teepee.timeout` is not the way to set timeout for teepee, we must set it when initialized

**Comment by pierreneter** [comment]
I'm trying
https://github.com/pierreneter-repositories/assetgraph/blob/webpack/lib/AssetGraph.js#L100
https://github.com/pierreneter-repositories/hyperlink/blob/webpack/lib/index.js#L31
https://github.com/pierreneter-repositories/hyperlink/blob/webpack/package.json#L42
https://github.com/webpack/webpack.js.org/blob/fix-hyperlink/package.json#L71

**Comment by skipjack** [comment]
> `teepee.timeout` is not the way to set timeout for teepee, we must set it when init

Yeah, that would make sense.

**Comment by montogeek** [comment]
@skipjack Is this PR still relevant?

**Comment by skipjack** [comment]
Yes and we should really try to get it resolved Bedford any big port.

**Comment by montogeek** [comment]
I will debug this, already read the entire thread, will ping @skipjack if any question arises

**Comment by montogeek** [comment]
I am afraid that our website graph is way too big to be resolved by assetgraph, this is a screenshot from a Memory profile
image

It takes a lot of time for assetgraph to solve relationships between assets in its internal implementation.

My node process have reached until 12 GB of memory processing up to 2166 assets, Travis limited is 3 GB for example.

One solution could be trying to exclude patterns from checking, like OpenColletive backers.

When running it along with https://github.com/mafintosh/why-is-node-running, it returns 14 handles, 12 of them from assetgraph:
<details>
```text
There are 14 handle(s) keeping the process running

# Timeout
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/hyperlink/lib/index.js:24  - setTimeout(function () {
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/hyperlink/bin/hyperlink:72 - var hyperlink = require('../lib/index');

# PROMISE
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/assetgraph/lib/transforms/loadAssets.js:10 - return async function loadAssets(assetGraph) {
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/assetgraph/lib/AssetGraph.js:502           - returnValue = transform(this);

# PROMISE
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/assetgraph/lib/transforms/loadAssets.js:11 - return _.flatten(await Promise.map(_.flatten(flattenedArguments), async assetConfig => {
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/assetgraph/lib/AssetGraph.js:502           - returnValue = transform(this);

# PROMISE
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/assetgraph/lib/transforms/loadAssets.js:11 - return _.flatten(await Promise.map(_.flatten(flattenedArguments), async assetConfig => {
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/assetgraph/lib/AssetGraph.js:502           - returnValue = transform(this);

# PROMISE
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/assetgraph/lib/AssetGraph.js:507 - returnValue.then(result => done(null, result), done);

# PROMISE
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/assetgraph/lib/transforms/loadAssets.js:11 - return _.flatten(await Promise.map(_.flatten(flattenedArguments), async assetConfig => {
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/util.js:16             - return target.apply(this, arguments);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/map.js:61              - var ret = tryCatch(callback).call(receiver, value, index, length);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/promise_array.js:114   - isResolved = this._promiseFulfilled(maybePromise, i);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/promise_array.js:78    - this._iterate(values);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/map.js:30              - this._init$(undefined, -2);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/async.js:138           - fn.call(receiver, arg);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/async.js:143           - this._drainQueue(this._normalQueue);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/async.js:17            - self._drainQueues();

# PROMISE
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/assetgraph/lib/assets/Asset.js:240         - async load() {
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/assetgraph/lib/transforms/loadAssets.js:24 - await asset.load();
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/util.js:16             - return target.apply(this, arguments);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/map.js:61              - var ret = tryCatch(callback).call(receiver, value, index, length);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/promise_array.js:114   - isResolved = this._promiseFulfilled(maybePromise, i);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/promise_array.js:78    - this._iterate(values);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/map.js:30              - this._init$(undefined, -2);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/async.js:138           - fn.call(receiver, arg);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/async.js:143           - this._drainQueue(this._normalQueue);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/async.js:17            - self._drainQueues();

# PROMISE
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/assetgraph/lib/assets/Asset.js:252         - this._rawSrc = await fs.readFileAsync(pathname);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/assetgraph/lib/transforms/loadAssets.js:24 - await asset.load();
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/util.js:16             - return target.apply(this, arguments);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/map.js:61              - var ret = tryCatch(callback).call(receiver, value, index, length);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/promise_array.js:114   - isResolved = this._promiseFulfilled(maybePromise, i);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/promise_array.js:78    - this._iterate(values);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/map.js:30              - this._init$(undefined, -2);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/async.js:138           - fn.call(receiver, arg);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/async.js:143           - this._drainQueue(this._normalQueue);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/async.js:17            - self._drainQueues();

# PROMISE
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/assetgraph/lib/assets/Asset.js:252         - this._rawSrc = await fs.readFileAsync(pathname);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/assetgraph/lib/transforms/loadAssets.js:24 - await asset.load();
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/util.js:16             - return target.apply(this, arguments);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/map.js:61              - var ret = tryCatch(callback).call(receiver, value, index, length);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/promise_array.js:114   - isResolved = this._promiseFulfilled(maybePromise, i);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/promise_array.js:78    - this._iterate(values);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/map.js:30              - this._init$(undefined, -2);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/async.js:138           - fn.call(receiver, arg);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/async.js:143           - this._drainQueue(this._normalQueue);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/async.js:17            - self._drainQueues();

# PROMISE
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/assetgraph/lib/transforms/loadAssets.js:24 - await asset.load();
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/util.js:16             - return target.apply(this, arguments);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/map.js:61              - var ret = tryCatch(callback).call(receiver, value, index, length);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/promise_array.js:114   - isResolved = this._promiseFulfilled(maybePromise, i);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/promise_array.js:78    - this._iterate(values);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/map.js:30              - this._init$(undefined, -2);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/async.js:138           - fn.call(receiver, arg);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/async.js:143           - this._drainQueue(this._normalQueue);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/async.js:17            - self._drainQueues();

# PROMISE
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/assetgraph/lib/transforms/loadAssets.js:24 - await asset.load();
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/util.js:16             - return target.apply(this, arguments);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/map.js:61              - var ret = tryCatch(callback).call(receiver, value, index, length);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/promise_array.js:114   - isResolved = this._promiseFulfilled(maybePromise, i);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/promise_array.js:78    - this._iterate(values);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/map.js:30              - this._init$(undefined, -2);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/async.js:138           - fn.call(receiver, arg);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/async.js:143           - this._drainQueue(this._normalQueue);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/async.js:17            - self._drainQueues();

# PROMISE
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/util.js:16           - return target.apply(this, arguments);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/thenables.js:63      - var result = util.tryCatch(then).call(x, resolve, reject);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/thenables.js:28      - return doThenable(obj, then, context);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/map.js:74            - var maybePromise = tryConvertToPromise(ret, this._promise);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/promise_array.js:114 - isResolved = this._promiseFulfilled(maybePromise, i);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/promise_array.js:78  - this._iterate(values);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/map.js:30            - this._init$(undefined, -2);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/async.js:138         - fn.call(receiver, arg);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/async.js:143         - this._drainQueue(this._normalQueue);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/bluebird/js/release/async.js:17          - self._drainQueues();

# PROMISE
(unknown stack trace)

# Immediate
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/assetgraph/lib/AssetGraph.js:499    - setImmediate(() => {
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/assetgraph/lib/TransformQueue.js:47 - that.assetGraph._runTransform(nextTransform, function (err, result) {
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/assetgraph/lib/TransformQueue.js:59 - that.run(cb, result);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/assetgraph/lib/AssetGraph.js:493    - cb(null, result);
/Users/montogeek/Infinite/webpack/webpack.js.org/node_modules/assetgraph/lib/AssetGraph.js:507    - returnValue.then(result => done(null, result), done);
```
</details>

**Comment by Munter** [comment]
Ugh. I thought we had succeeded in unloading assets that were already visited and handled :(

@papandreou Do you think it's possible to reduce the memory usage drastically for the hyperlink use case, or would hyperlink have to go more low level and manually do the asset traversal without the complete graph?

**Comment by montogeek** [comment]
@Munter You mean this https://github.com/Munter/hyperlink/blob/cf7dba494ca8c3c0ffba757a60d1342ccbb6c336/lib/index.js#L287-L301?

**Comment by papandreou** [comment]
@Munter, it'd be great to be able to "page out" assets while keeping their outgoing relations, but we aren't there yet.

Let's upgrade to ag 4 and take another stab at the automatic unloading.

**Comment by Munter** [comment]
I'm not sure what's going on with Travis here. It's telling me the branch is not built per configuration and I don't know how to override to make it run manually

image

**Comment by montogeek** [comment]
I cant find a way to trigger it manually, Can you push an empty commit to a JS file?

**Comment by montogeek** [comment]
@Munter Solving conflicts triggers a new build

**Comment by montogeek** [comment]
@Munter Node 6 doesn't support `async` :/ We should update it to latest LTS

**Comment by Munter** [comment]
Ah, yes. Assetgraph 4 has a node 8 minimum version. I don't know if that's too much of an imposition on the webpack docs project

**Comment by Munter** [comment]
I upped Travis to node 8, just to see if this even works at all. Then you can make an informed decision

**Comment by montogeek** [comment]
I think it will be just fine

**Comment by montogeek** [comment]
@Munter Node 8 works.
It still runs out of memory with new assetgraph

**Comment by Munter** [comment]
Damn. We've had successful runs on some of our laptops on a local docs build without tweaking node's available memory. On my personal laptop it still fails on a memory problem, but I was hoping my personal setup was not representative of Travis :(

**Comment by montogeek** [comment]
Travis limit is 3 GB, our local computers probably have bigger memory

**Comment by montogeek** [comment]
We could attach an Inspector to it and see what is burning memory

**Comment by montogeek** [comment]
@Munter It was doing just fine, some links are 404'd :/

**Comment by Munter** [comment]
I had it pass on my otherwise failing local laptop. I think we got the memory issue. It wasn't supposed to throw on missing files though. I'm trying to recreate that travis error locally to fix it

**Comment by Munter** [comment]
Can't recreate that travis failure on a fresh `master` docs build and the tip of that hyperlink branch :(

File: package.json [hunk 1]
```diff
@@ -32,7 +32,7 @@
     "fetch": "sh src/scripts/fetch.sh",
     "init:generated": "mkdirp ./generated/loaders && mkdirp ./generated/plugins ",
     "lint": "run-s lint:*",
-    "lint:links": "hyperlink -r build/index.html --canonicalroot https://webpack.js.org/ --debug | node ./src/scripts/check-links.js",
+    "lint:links": "hyperlink -r build/index.html --canonicalroot https://webpack.js.org/",
```
Comment (montogeek) [inline package.json]: :)

Comment (Munter) [reply package.json]: I just can't recreate this damn failure :'(

**Comment by montogeek** [comment]
It works!

**Comment by Munter** [comment]
Seems like there is something in `check-links.js` that doesn't handle the hyperlink output correctly. This should be possible to recreate by just piping the saved output from hyperlink directly into it

**Comment by Munter** [comment]
We'll do some work on allowing a programmatic interface to define filter that can mark a failed test as `todo` or `skip`. This should allow programmatic usage in the webpack docs without the `check-links` script. To reduce the verbose ouput the resulting tap stream can just be piped through a minimal tap reporter that strips away succeeding tests

**Comment by Munter** [comment]
@montogeek @skipjack The link check is running successfully now. How you you like the new output? https://travis-ci.org/webpack/webpack.js.org/builds/360812190#L1218

**Comment by skipjack** [comment]
Awesome @Munter, the output looks great! @montogeek wanna look into the broken links? Once those are fixed we should get this into `master` and then rebase or merge to `rebuild` to pick this up there. Having this back in is clutch for a big port.

**Comment by Munter** [comment]
We're still running against a github hash on this branch, so I still need to clean this whole thing up. We have basically rewritten the entirety of hyperlink with all the work done over easter, so it's a major release

**Comment by skipjack** [comment]
No problem, we can hold off on the merge until you push the major or pre-release and we pick it up over here.

**Comment by montogeek** [comment]
I am on it, we already fixed some on `master`, but looks like there is more

**Comment by Munter** [comment]
We probably also need to fine tune a bit on what is reported as a blocking error, like loading internal assets, and what can be deemed acceptable for an immediate PR.

For example a lot of external urls will close down, get moved or have unstable servers. That might fail the checks for redirects, which might not be related to your immediate needs on an unrelated PR.

This find tuning can be done by using the new `--todo` switch and match on the types of tests that you want to accepts a failure of, but still have reported. In the best case some sort of feedback loop into a github issue or slack notification could be triggered

**Comment by Munter** [comment]
Since hyperlink is set up in a way right now where it will fail the build, I think we need to do a few tweaks to reduce either the scope of the tests or the severity.

This version of hyperlink has a `--todo` option, which allows you to define matching patterns for test that are acceptable failures. We should probably use that to allow people to create PR's that are not affected by the outside world.

I see a few different ways of going about this:

## Mark test of imported content as `todo`

We are fetching a bunch of content from the readme of a lot of plugins. If there are errors in any of those it will affect the build. But the individual plugin doesn't have any way of anticipating that their change will break the docs build. If we can identify certain routes that purely contain external content, we could mark the tests of outgoing links from those pages as `todo`.

## Skip tests of external links completely on a PR test

The most important piece of information hyperlink gives you is if your internally controlled website is an intact dependency graph. We could reduce the amount of checks done on a normal PR to only look at internal structure.

This should probably be paired with a periodical test that runs the full hyperlink suite, so the docs site isn't pointing at external missing pages. I'm not aware of any really awesome reporting tools that such an error output could be posted to in a structured way. I think you can do slack integrations with travis, so maybe there should be a daily full check with errors reported to a slack channel for the community to pick up along the way without them being build blockers

**Comment by montogeek** [comment]
@Munter Sounds good

**Comment by Munter** [comment]
The change set was getting a bit unwieldy, so I squashed everything to keep this branch manageable. The change set is quite small after all the experiments. No need to keep that history

**Comment by montogeek** [comment]
@Munter What else is missing to merge this one?

**Comment by Munter** [comment]
@montogeek I can't build the current master on my machine, so I can't iterate. I need the updated rebuild branch merged before I can continue

**Comment by montogeek** [comment]
@Munter Why it is not building for you? :/

**Comment by Munter** [comment]
I can't wait 50 minutes for a build, which is what the current master takes on my MacBook air

**Comment by montogeek** [comment]
@Munter Would it help if I send you a zip file with `build` dir content?

**Comment by Munter** [comment]
Not really. I need the ability to iterate and fix things as I go

**Comment by skipjack** [comment]
@Munter if that's the case, then maybe rebase on `rebuild` and repoint this PR to `rebuild`? It's getting pretty close to done, just working on getting the last few PRs in and then we'll just have to do some thorough testing to make sure we didn't miss anything.

**Comment by skipjack** [comment]
I'm still strapped on time but would love to see this finished. If anyone has bandwidth to dig into it before I do -- which might still be a few months out -- I highly recommend rebasing on `rebuild`, re-pointing this PR to `rebuild`, and then testing to see what else is needed to call this complete.

**Comment by montogeek** [comment]
I could do it tomorrow

**Comment by Munter** [comment]
CLosing in favor of https://github.com/webpack/webpack.js.org/pull/2382

**Pull request closed without merging**